src/nvme/nbft.c

```diff
@@ -0,0 +1,947 @@
+			"invalid security profile desciptor list offset");
+		raw_security_array = (struct raw_nbft_security *)(raw_nbft + control->security_profile_descriptor_list_offset);
+		read_security_descriptors(nbft, control->num_sec, raw_security_array, control->security_profile_descriptor_length);
+	}
+
+	/*
+	 * discovery
+	 */
+	if (control->num_disc > 0) {
+		struct raw_nbft_discovery *raw_discovery_array;
+
+		verify(control->discovery_profile_descriptor_list_offset + control->discovery_profile_descriptor_length * control->num_disc <= header->length,
+		       "invalid discovery profile descriptor list offset");
+		raw_discovery_array = (struct raw_nbft_discovery *)(raw_nbft + control->discovery_profile_descriptor_list_offset);
+		read_discovery_descriptors(nbft, control->num_disc, raw_discovery_array, control->discovery_profile_descriptor_length);
+	}
+
+	/*
+	 * subsystem namespace
+	 */
+	if (control->num_ssns > 0) {
+		struct raw_nbft_ssns *raw_ssns_array;
+
+		verify(control->ssns_descriptor_list_offset + control->ssns_descriptor_length * control->num_ssns <= header->length,
+		       "invalid subsystem namespace descriptor list offset");
+		raw_ssns_array = (struct raw_nbft_ssns *)(raw_nbft + control->ssns_descriptor_list_offset);
+		read_ssns_descriptors(nbft, control->num_ssns, raw_ssns_array, control->ssns_descriptor_length);
+	}
+
+	return 0;
+}
+
+void nbft_free(struct nbft_info *nbft)
+{
+	void *subtable;
+	struct nbft_subsystem_ns *ns;
+
+	while ((subtable = list_pop(&nbft->hfi_list, struct nbft_hfi, node)))
+		free(subtable);
+	while ((subtable = list_pop(&nbft->discovery_list, struct nbft_discovery, node)))
+		free(subtable);
+	while ((subtable = list_pop(&nbft->security_list, struct nbft_security, node)))
+		free(subtable);
+	while ((ns = list_pop(&nbft->subsystem_ns_list, struct nbft_subsystem_ns, node))) {
+		free(ns->hfis);
+		free(ns);
+	}
+	free(nbft->raw_nbft);
+	free((void *)nbft->filename);
+	free(nbft);
+}
+
+/**
+ * nbft_read - read ACPI NBFT table and parse contents into struct nbft_info 
+ * @nbft: will contain address of struct nbft_info if read successful
+ * @filename: location of raw ACPI NBFT table
+ *
+ * Returns 0 on success, errno otherwise.
+ */
+int nbft_read(struct nbft_info **nbft, const char *filename)
+{
+	__u8 *raw_nbft = NULL;
+	size_t raw_nbft_size;
+	FILE *raw_nbft_fp;
+	int i, ret = 0;
+
+	/*
+	 * read in raw nbft file
+	 */
+	raw_nbft_fp = fopen(filename, "rb");
+	if (raw_nbft_fp == NULL) {
+		nvme_msg(NULL, LOG_ERR, "Failed to open %s: %s\n", filename, strerror(errno));
+		return -EINVAL;
+	}
+
+	i = fseek(raw_nbft_fp, 0L, SEEK_END);
+	if (i) {
+		nvme_msg(NULL, LOG_ERR, "Failed to read from %s: %s\n", filename, strerror(errno));
+		ret = -EINVAL;
+		goto fail_1;
+	}
+
+	raw_nbft_size = ftell(raw_nbft_fp);
+	rewind(raw_nbft_fp);
+
+	raw_nbft = malloc(raw_nbft_size);
+	if (!raw_nbft) {
+		nvme_msg(NULL, LOG_ERR, "Failed to allocate memory for NBFT table");
+		ret = -ENOMEM;
+		goto fail_1;
```

Comment:
https://www.kernel.org/doc/html/v4.10/process/coding-style.html#centralized-exiting-of-functions:

> Choose label names which say what the goto does or why the goto exists. An example of a good name could be out_free_buffer: if the goto frees buffer. Avoid using GW-BASIC names like err1: and err2:, as you would have to renumber them if you ever add or remove exit paths, and they make correctness difficult to verify anyway.

Reply:
Refactored.